weio/fast_output_file.py

```diff
@@ -113,52 +125,92 @@ def readline(iLine):
+                        self['binary']=False
+            except MemoryError as e:
+                raise BrokenReaderError('FAST Out File {}: Memory error encountered\n{}'.format(self.filename,e))
+            except Exception as e:
+                raise WrongFormatError('FAST Out File {}: {}'.format(self.filename,e.args))
+            if self.data.shape[0]==0:
+                raise EmptyFileError('This FAST output file contains no data: {}'.format(self.filename))
+
+
+
+        # --- Store header information
+        self['attribute_names'] = info.get('attribute_names', [])
+        self['attribute_units'] = info.get('attribute_units', [])
         self.description = info.get('description', '')
         self.description = ''.join(self.description) if isinstance(self.description,list) else self.description
-        if isinstance(self.data, pd.DataFrame):
-            self.data.columns = cols
-        else:
-            if len(cols)!=self.data.shape[1]:
-                raise BrokenFormatError('Inconstistent number of columns between headers ({}) and data ({}) for file {}'.format(len(cols), self.data.shape[1], self.filename))
-            self.data = pd.DataFrame(data=self.data, columns=cols)
+
+        # Store binary file metadata for streaming mode
+        if self.streaming and self['binary']:
+            self['_FileID'] = info.get('FileID')
+            self['_NumOutChans'] = info.get('NumOutChans')
+            self['_NT'] = info.get('NT')
+            self['_ColScl'] = info.get('ColScl')
+            self['_ColOff'] = info.get('ColOff')
+            if info['FileID'] == FileFmtID_WithTime:
+                self['_TimeScl'] = info.get('TimeScl')
+                self['_TimeOff'] = info.get('TimeOff')
+            else:
+                self['_TimeOut1'] = info.get('TimeOut1')
+                self['_TimeIncr'] = info.get('TimeIncr')
+
+        # --- Convert to DataFrame (only if data was loaded)
+        if self.data is not None:
+            if info['attribute_units'] is not None:
+                info['attribute_units'] = [re.sub(r'[()\[\]]','',u) for u in info['attribute_units']]
+                if len(info['attribute_names'])!=len(info['attribute_units']):
+                    cols=info['attribute_names']
+                    print('[WARN] not all columns have units! Skipping units')
+                else:
+                    cols=[n+'_['+u.replace('sec','s')+']' for n,u in zip(info['attribute_names'], info['attribute_units'])]
+            else:
+                cols=info['attribute_names']
+            if isinstance(self.data, pd.DataFrame):
+                self.data.columns = cols
+            else:
+                if len(cols)!=self.data.shape[1]:
+                    raise BrokenFormatError('Inconstistent number of columns between headers ({}) and data ({}) for file {}'.format(len(cols), self.data.shape[1], self.filename))
+                self.data = pd.DataFrame(data=self.data, columns=cols)
 

```

Comment:
The lines above that "convert to dataframe" are now repeated 3 times in this file.
Let's make a function hawc2data_2_dataframe, that take the data, the info and the filename (to throw an exception),  and returns a dataframe, and the info (because the units are cleaned up by the function)